src/retry_logic.rs

```diff
@@ -0,0 +1,110 @@
+use std::time::Duration;
+use temporal_sdk_core_protos::coresdk::common::RetryPolicy;
+
+pub(crate) trait RetryPolicyExt {
+    /// Ask this retry policy if a retry should be performed. Caller provides the current attempt
+    /// number - the first attempt should start at 1.
+    ///
+    /// Returns `None` if it should not, otherwise a duration indicating how long to wait before
+    /// performing the retry.
+    fn should_retry(&self, attempt_number: usize, err_str: &str) -> Option<Duration>;
+}
+
+impl RetryPolicyExt for RetryPolicy {
+    fn should_retry(&self, attempt_number: usize, err_str: &str) -> Option<Duration> {
+        if attempt_number >= self.maximum_attempts.max(0) as usize {
+            return None;
+        }
+
+        for pat in &self.non_retryable_error_types {
+            if err_str.contains(pat.as_str()) {
```

Comment:
So I think that approach kinda sucks and we should ideally change it on server at some point too. It's deeply inflexible.

It's even worse in this context because I have no idea what type name you're talking about and the user likely doesn't either. The type names they see in lang I won't even see in Core. It just doesn't make sense. We would have to add an explicit field in `Failure` (which is upstream, not owned by us) along the lines of `failure_type_string` just to be able to check this reliably.

I'd much rather make it clear that for local activities it works this way, which most users I think will prefer in any case.

Reply:
> We would have to add an explicit field in Failure (which is upstream, not owned by us) along the lines of failure_type_string just to be able to check this reliably.

:+1: If we don't want to be limited to a specific set of types but also don't want to do gross error message string-contains, we'd need some kind of "tagging".

> I'd much rather make it clear that for local activities it works this way, which most users I think will prefer in any case.

I do think `non_retryable_error_types` is specified by https://github.com/temporalio/api/blob/master/temporal/api/common/v1/message.proto#L109-L111 to mean failure type for now and I worry that someone may reuse a retry policy for normal activities and local activities expecting this field to work the same across both. Can a language not provide the failure types of their failures to core for purposes of this kind of matching?

Reply:
@bergundy I didn't see the `type` field on `ApplicationFailure`. I will only use/look for that which is seemingly closest to normal activities.

I still think it's really difficult for users to make use of this well, but better to be consistent I suppose